File: framework/src/play-guice/src/main/java/play/libs/akka/AkkaGuiceSupport.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package play.libs.akka;

import akka.actor.Actor;
import akka.actor.ActorRef;
import akka.actor.Props;
import com.google.inject.assistedinject.FactoryModuleBuilder;
import com.google.inject.name.Names;
import com.google.inject.util.Providers;
import play.libs.Akka;

import java.util.function.Function;

/**
 * Support for binding actors with Guice.
 *
 * Mix this interface in with a Guice AbstractModule to get convenient support for binding actors.  For example:
 * <pre>
 * public class MyModule extends AbstractModule implements AkkaGuiceSupport {
 *   protected void configure() {
 *     bindActor(MyActor.class, "myActor");
 *   }
 * }
 * </pre>
 *
 * Then to use the above actor in your application, add a qualified injected dependency, like so:
 * <pre>
 * public class MyController extends Controller {
 *   {@literal @}Inject @Named("myActor") ActorRef myActor;
 *   ...
 * }
 * </pre>
 */
public interface AkkaGuiceSupport {

    /**
     * Bind an actor.
     *
     * This will cause the actor to be instantiated by Guice, allowing it to be dependency injected itself.  It will
     * bind the returned ActorRef for the actor will be bound, qualified with the passed in name, so that it can be
     * injected into other components.
     *
     * @param <T> the actor type.
     * @param actorClass The class that implements the actor.
     * @param name The name of the actor.
     * @param props A function to provide props for the actor. The props passed in will just describe how to create the
     *              actor, this function can be used to provide additional configuration such as router and dispatcher
     *              configuration.
     */
    default <T extends Actor> void bindActor(Class<T> actorClass, String name, Function<Props, Props> props) {
        BinderAccessor.binder(this).bind(ActorRef.class)
                .annotatedWith(Names.named(name))
                .toProvider(Providers.guicify(Akka.providerOf(actorClass, name, props)))
                .asEagerSingleton();
    }

    /**
     * Bind an actor.
     *
     * This will cause the actor to be instantiated by Guice, allowing it to be dependency injected itself.  It will
     * bind the returned ActorRef for the actor will be bound, qualified with the passed in name, so that it can be
     * injected into other components.
     *
     * @param <T> the actor type.
     * @param actorClass The class that implements the actor.
     * @param name The name of the actor.
     */
    default <T extends Actor> void bindActor(Class<T> actorClass, String name) {
        bindActor(actorClass, name, Function.identity());
    }

    /**
     * Bind an actor factory.
     *
     * This is useful for when you want to have child actors injected, and want to pass parameters into them, as well as
     * have Guice provide some of the parameters.  It is intended to be used with Guice's AssistedInject feature.
     *
     * Let's say you have an actor that looks like this:
     *
     * <pre>
     * public class MyChildActor extends AbstractActor {
     *     final Database db;
     *     final String id;
     *
     *     {@literal @}Inject
     *     public MyChildActor(Database db, {@literal @}Assisted String id) {
     *         this.db = db;
     *         this.id = id;
     *     }
     *     ...
     * }
     * </pre>
     *
     * So {@code db} should be injected, while {@code id} should be passed.  Now, define an interface that takes the id,
     * and returns the actor:
     *
     * <pre>
     * public interface MyChildActorFactory {
     *   MyChildActor apply(String id);
     * }
     * </pre>
     *
     * Now you can use this method to bind the child actor in your module:
     *
     * <pre>
     * public class MyModule extends AbstractModule implements AkkaGuiceSupport {
     *   protected void configure() {
     *     bindActorFactory(MyChildActor.class, MyChildActorFactory.class);
     *   }
     * }
     * </pre>
     *
     * Now, when you want an actor to instantiate this as a child actor, inject `MyChildActorFactory`:
     *
     * <pre>
     * public class MyActor extends AbstractActor implements InjectedActorSupport {
     *   final MyChildActorFactory myChildActorFactory;
     *
     *   {@literal @}Inject
     *   public MyActor(MyChildActor myChildActorFactory) {
     *       this.myChildActorFactory = myChildActorFactory;
     *   }
     *
     *   {@literal @}Override
     *   public Receive createReceive() {
     *   return receiveBuilder()
     *     .match(CreateChildActor.class, msg -&gt; {
     *       ActorRef child = injectedChild(myChildActorFactory.apply(msg.getId()));
     *       sender().send(child, self);
     *     }
     *     .build()
     *   }
     * }
     * </pre>
     *
     * @param <T> the actor type.
     * @param actorClass The class that implements the actor.
     * @param factoryClass The factory interface for creating the actor.
     */
    default <T extends Actor> void bindActorFactory(Class<T> actorClass, Class<?> factoryClass) {
        BinderAccessor.binder(this).install(
                new FactoryModuleBuilder()
                        .implement(Actor.class, actorClass)
                        .build(factoryClass)
        );
    }
}

File: framework/src/play-guice/src/main/scala/play/api/libs/concurrent/AkkaGuiceSupport.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.concurrent

import java.lang.reflect.Method

import akka.actor._
import com.google.inject._
import com.google.inject.assistedinject.FactoryModuleBuilder

import scala.reflect._

/**
 * Support for binding actors with Guice.
 *
 * Mix this trait in with a Guice AbstractModule to get convenient support for binding actors.  For example:
 * {{{
 *   class MyModule extends AbstractModule with AkkaGuiceSupport {
 *     def configure = {
 *       bindActor[MyActor]("myActor")
 *     }
 *   }
 * }}}
 *
 * Then to use the above actor in your application, add a qualified injected dependency, like so:
 * {{{
 *   class MyController @Inject() (@Named("myActor") myActor: ActorRef, val controllerComponents: ControllerComponents)
 *       extends BaseController {
 *     ...
 *   }
 * }}}
 */
trait AkkaGuiceSupport {
  self: AbstractModule =>

  import com.google.inject.name.Names
  import com.google.inject.util.Providers

  private def accessBinder: Binder = {
    val method: Method = classOf[AbstractModule].getDeclaredMethod("binder")
    if (!method.isAccessible) {
      method.setAccessible(true)
    }
    method.invoke(this).asInstanceOf[Binder]
  }

  /**
   * Bind an actor.
   *
   * This will cause the actor to be instantiated by Guice, allowing it to be dependency injected itself.  It will
   * bind the returned ActorRef for the actor will be bound, qualified with the passed in name, so that it can be
   * injected into other components.
   *
   * @param name The name of the actor.
   * @param props A function to provide props for the actor. The props passed in will just describe how to create the
   *              actor, this function can be used to provide additional configuration such as router and dispatcher
   *              configuration.
   * @tparam T The class that implements the actor.
   */
  def bindActor[T <: Actor: ClassTag](name: String, props: Props => Props = identity): Unit = {
    accessBinder.bind(classOf[ActorRef])
      .annotatedWith(Names.named(name))
      .toProvider(Providers.guicify(Akka.providerOf[T](name, props)))
      .asEagerSingleton()
  }

  /**
   * Bind an actor factory.
   *
   * This is useful for when you want to have child actors injected, and want to pass parameters into them, as well as
   * have Guice provide some of the parameters.  It is intended to be used with Guice's AssistedInject feature.
   *
   * Let's say you have an actor that looks like this:
   *
   * {{{
   * class MyChildActor @Inject() (db: Database, @Assisted id: String) extends Actor {
   *   ...
   * }
   * }}}
   *
   * So `db` should be injected, while `id` should be passed.  Now, define a trait that takes the id, and returns
   * the actor:
   *
   * {{{
   * trait MyChildActorFactory {
   *   def apply(id: String): Actor
   * }
   * }}}
   *
   * Now you can use this method to bind the child actor in your module:
   *
   * {{{
   *   class MyModule extends AbstractModule with AkkaGuiceSupport {
   *     def configure = {
   *       bindActorFactory[MyChildActor, MyChildActorFactory]
   *     }
   *   }
   * }}}
   *
   * Now, when you want an actor to instantiate this as a child actor, inject `MyChildActorFactory`:
   *
   * {{{
   * class MyActor @Inject() (myChildActorFactory: MyChildActorFactory) extends Actor with InjectedActorSupport {
   *
   *   def receive {
   *     case CreateChildActor(id) =>
   *       val child: ActorRef = injectedChild(myChildActorFactory(id), id)
   *       sender() ! child
   *   }
   * }
   * }}}
   *
   * @tparam ActorClass The class that implements the actor that the factory creates
   * @tparam FactoryClass The class of the actor factory
   */
  def bindActorFactory[ActorClass <: Actor: ClassTag, FactoryClass: ClassTag]: Unit = {
    accessBinder.install(new FactoryModuleBuilder()
      .implement(classOf[Actor], implicitly[ClassTag[ActorClass]].runtimeClass.asInstanceOf[Class[_ <: Actor]])
      .build(implicitly[ClassTag[FactoryClass]].runtimeClass))
  }

}

